source/vir/src/sst_to_air.rs

```diff
@@ -1912,43 +1923,40 @@ fn stm_to_stmts(ctx: &Ctx, state: &mut State, stm: &Stm) -> Result<Vec<Stmt>, Vi
             let mut stmts: Vec<Stmt> = Vec::new();
-            if let Some(x) = loc_is_var(dest) {
-                let name = suffix_local_unique_id(x);
-                stmts.push(Arc::new(StmtX::Assign(name, exp_to_expr(ctx, rhs, expr_ctxt)?)));
-                if ctx.debug {
-                    // Add a snapshot after we modify the destination
-                    let sid = state.update_current_sid(SUFFIX_SNAP_MUT);
-                    let snapshot = Arc::new(StmtX::Snapshot(sid.clone()));
-                    stmts.push(snapshot);
-                    // Update the snap_map so that it reflects the state _after_ the
-                    // statement takes effect.
-                    state.map_span(&stm, SpanKind::Full);
-                }
-            } else {
-                let (base_var, LocFieldInfo { base_typ, base_span, a: fields }) =
-                    loc_to_field_path(dest);
-                stmts.push(Arc::new(StmtX::Snapshot(snapshot_ident(SNAPSHOT_ASSIGN))));
-                stmts.push(Arc::new(StmtX::Havoc(suffix_local_unique_id(&base_var))));
-                let snapshotted_rhs = snapshotted_vars(rhs, SNAPSHOT_ASSIGN);
-                let eqx = ExpX::Binary(BinaryOp::Eq(Mode::Spec), dest.clone(), snapshotted_rhs);
-                let eq = SpannedTyped::new(&stm.span, &Arc::new(TypX::Bool), eqx);
-                stmts.extend(stm_to_stmts(
-                    ctx,
-                    state,
-                    &Spanned::new(stm.span.clone(), StmX::Assume(eq)),
-                )?);
-                stmts.extend(assume_other_fields_unchanged(
-                    ctx,
-                    SNAPSHOT_ASSIGN,
-                    &stm.span,
-                    &base_var,
-                    &LocFieldInfo { base_typ, base_span, a: vec![fields] },
-                    expr_ctxt,
-                )?);
-                if ctx.debug {
-                    unimplemented!("complex assignments are unsupported in debugger mode");
+            if ctx.debug {
+                unimplemented!("assignments are unsupported in debugger mode");
+            }
+
+            let dest = dest;
+            let mut value = exp_to_expr(ctx, &rhs, expr_ctxt)?;
+            let mut value_typ = rhs.typ.clone();
+
+            let (base_var, LocFieldInfo { base_typ, base_span: _, a: fields }) =
+                loc_to_field_update_data(&dest);
+
+            for FieldUpdateDatum { opr, field_typ, base_exp } in fields.iter().rev() {
+                let acc = variant_field_ident_internal(
+                    &encode_dt_as_path(&opr.datatype),
+                    &opr.variant,
+                    &opr.field,
+                    true,
+                );
+                let bop = air::ast::BinaryOp::FieldUpdate(acc);
+                if typ_is_poly(ctx, field_typ) && !typ_is_poly(ctx, &value_typ) {
+                    value = try_box(ctx, value, &value_typ).expect("box field update");
```

Comment:
If possible, I'd like to see boxing/unboxing coercions inserted by poly.rs rather than by sst_to_air.rs.  The ongoing `monomorphize` branch, for example, would have to add special cases to sst_to_air.rs whenever sst_to_air.rs performs boxing/unboxing operations directly, whereas everything in poly.rs will automatically be skipped by `monomorphize`.